### DISCUSSION TITLE
SDL2 and threads?

### DISCUSSION THREAD

**Discussion by gregbuchholz** [emscripten-core/emscripten#16212] [Q&A]
Just wondering what the latest thinking on running SDL2 and threads with emscripten and asyncify and proxy-to-pthread.  I have a simple rust [example program](https://github.com/gregbuchholz/wasm_sdl_threads) that fails at runtime with the following exception:
```
worker.js onmessage() captured an uncaught exception: TypeError: SDL2.ctx is undefined
```
...that message occurs after the first trip through the event loop, so everything has been initialized "successfully".  The program is just compiled with pthread support, it doesn't actually spawn any additional threads.  So I'm wondering if it has something to do with the proxy-to-pthread option.  My ultimate goal is to use threads for accelerating a complex computation, not to interact with SDL in a threaded manner.

Options currently in use:

```
-pthread -sUSE_SDL=2 -s PROXY_TO_PTHREAD -s ASYNCIFY -s ALLOW_MEMORY_GROWTH=1 -s USE_PTHREADS=1 -s PTHREAD_POOL_SIZE=8 -s OFFSCREEN_FRAMEBUFFER=1
```
...it seems like there has been some work on SDL and threads previously: [ [WIP] SDL2 thread proxying fixes #127](https://github.com/emscripten-ports/SDL2/pull/127).

Thanks!

**Comment by gregbuchholz** [comment]
There is also a version showing the same problem using C instead of Rust.  That code is over at [c_example/](https://github.com/gregbuchholz/wasm_sdl_threads/tree/main/c_example).  See the [Makefile](https://github.com/gregbuchholz/wasm_sdl_threads/blob/main/c_example/Makefile) for various targets, and try running `make run` to see the failing case, and `make run-single` for the successful case.  The only difference is the arguments passed to `emcc`, with the failing one getting:
```
emcc t_test.c -o t_test.html -pthread -sUSE_SDL=2 -s PROXY_TO_PTHREAD -s ASYNCIFY -s ALLOW_MEMORY_GROWTH=1 -s USE_PTHREADS=1 -s PTHREAD_POOL_SIZE=8 -s OFFSCREEN_FRAMEBUFFER=1
```

...and the [error message](https://github.com/gregbuchholz/wasm_sdl_threads/blob/main/c_example/exception_with_c.txt):
```
worker.js onmessage() captured an uncaught exception: TypeError: GLctx is undefined
```

**Comment by kripken** [comment]
I am not sure if that PR you linked (https://github.com/emscripten-ports/SDL2/pull/127) has landed in upstream SDL - if it has, then it should be included already since we switched to upstream.

@Daft-Freak do you know?

If we still need those fixes then they'd need to be done in upstream now.

**Reply by Daft-Freak** [reply]
I think the "use new APIs" bits landed, but the rest hasn't.
